### PR TITLE
Fixes Refresh Error

### PR DESCRIPTION
**Pull request by edjiang**
Currently, ID Site has a bug where if you load the login page, and refresh, you get the following error: 

screen shot 2016-05-05 at 11 07 20 pm

This is very easy to replicate. Go to any ID Site (you can use Tom's demo at http://shielded-journey-8142.herokuapp.com by example) and you can replicate it.

I've submitted a fix for this bug. 

This passes all tests but one, so I will need @robertjd, @rdegges  or someone more familiar with JS to help me with the test suite. 

```
    describe('without a JWT in the url, and not specified by options', function () {
      var result;
      before(function(done){
        new stormpathJs.Client(function(err,res){
          result = [err,res];
          done();
        });
      });
      it('should err',function(){
        assert.equal(result[0].message,strings.errors.JWT_NOT_FOUND);
      });
    });
```

This test fails as I'm using the window object directly, and I'm not quite sure the best strategy to mock the window object. Can one of you help me fix that test, merge, and tag? Once we fix this, we should update stormpath/idsite-src and stormpath/idsite to fix this bug.

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -132,6 +135,14 @@ Client.prototype.getJwtFromUrl = function () {
 };
 
 /**
+ * Attempts to replace the JWT in the url with the new JWT passed in the function.
+ * @param {string} newJwt
+ */
+Client.prototype.replaceJWTInUrl = function (newJwt) {
+  window.location.hash = '/?jwt=' + newJwt;
```
Comment (timothyej) [inline lib/client.js]: The original JWT is extracted from `window.location.href`, which doesn't necessarily have to be the hash, even though that is what ID Site is using. However, updating other parts of the URL other than the hash will cause it to reload and then the new JWT will be consumed again, so I'm not sure how to solve this problem.

File: lib/client.js [hunk 2]
```diff
@@ -80,6 +80,9 @@ function Client (options,readyCallback) {
       if (!self.requestExecutor.authToken) {
         return cb(new Error(strings.errors.NO_AUTH_TOKEN_HEADER));
       }
+
+      self.replaceJWTInUrl(self.requestExecutor.authToken);
```
Comment (timothyej) [inline lib/client.js]:
```javascript
27    self.jwt = opts.token || self.getJwtFromUrl();
```
If the original JWT comes from `opts` rather than the URL, I don't think it should store the new JWT in the URL? :)

Comment (edjiang) [inline lib/client.js]: That makes sense. I'll try that.

File: lib/client.js (hunk 1 above)
Comment (edjiang) [inline lib/client.js]: We don't use the query string for this right now, and I don't think we should be putting the JWT in the query string at all anyways, for the same reasons as social login, so I would consider this a minor issue. We would know if we were changing how stormpath calls into ID Site, anyways.

Comment (robertjd) [inline lib/client.js]: The jwt in the URL has to be obtained from the /sso endpoint during
redirect to the ID site.  It is one time use and cannot be used again, so
this patch will not work.

Comment (edjiang) [inline lib/client.js]: This patch will work, I've tested it. You already have most of the code for this. When getting the ID Site view model, stormpath's response view models contains an authorization header that contains a new JWT. You are grabbing it and setting it to the authToken variable in your code. We just need to persist this somehow so on refresh that token gets used.

Comment (timothyej) [inline lib/client.js]: @edjiang @robertjd I've also verified that it works 👍

Comment (timothyej) [inline lib/client.js]: @edjiang Oh, I didn't know this was only used for ID Site. Then I guess it's fine :)

Comment (robertjd) [inline lib/client.js]: The secondary jwt is only good for five minutes, there are other edge cases
to consider here.

Comment (edjiang) [inline lib/client.js]: I will change the error message so it's more human friendly as well then.

Comment (edjiang) [inline lib/client.js]: I've fixed the test problem, and edited the error messages. Good to merge & `grunt release`? 

By the way, Contributing.md is super useful; I am not a frontend js guy so it helped me get the toolchain set up really easily =]

**Comment by robertjd** [comment]
I'm assigning this to myself for detailed review, might not happen today though.

**Comment by edjiang** [comment]
OK, great. If replacing the URL is not an option, we can always use cookies or local storage for this too.

**Comment by robertjd** [comment]
I'm finally looking at this now, thanks @edjiang for the solution!  This does seem to work for some of the user stories, but I did run into one issue just now: after triggering a password reset link to yourself, and then clicking on the link, you arrive at the login form (not the reset form) because the entire URL gets overwritten and the #reset hash is lost.  

There may be other cases that are affected, but I'm only able to get this far today.

**Comment by edjiang** [comment]
OK, so I took a look at this further yesterday and tried both string replacing and switching to a cookie based approach. 

I think the cookie based approach is more ideal since otherwise if you navigate between the login/registration page, that updates the url and the token is lost when refreshing. 

However, ID Site is currently looking for the reset token in one spot. So I added an additional location for it to check for the password reset token. 

This works when I play around with the site, but currently Protractor.js doesn't seem to pass on `idSite-src`. However it seems to not pass for the current version of master, so I'm not sure if it's my config.

Ready for review @robertjd

File: lib/client.js [hunk 3]
```diff
@@ -124,13 +129,62 @@ Client.prototype.setCachedOrganizationNameKey = function (nameKey) {
...
+ * @return {string} Password Reset Token
+ */
+Client.prototype.getPasswordResetToken = function() {
+  var self = this;
+  if (self.jwtPayload.sp_token) {
+    return self.jwtPayload.sp_token;
+  } else {
+    // We need to get the token from jwt.scope.application.SLUG.passwordResetToken.SLUG
+    // See https://gist.github.com/edjiang/a570a4d9e60d08492def97938b35cdfa for an example token
+    var appSlug = self.appHref.replace('https://api.stormpath.com/v1/applications/', '');
```
Comment (robertjd) [inline lib/client.js]: "appSlug" -> "applicationId", would be more descriptive

**Comment by robertjd** [comment]
Thanks @edjiang for the solution!  I'm preparing the release now.
